core/src/main/java/org/apache/spark/shuffle/api/WritableByteChannelWrapper.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api;
+
+import java.io.Closeable;
+import java.nio.channels.WritableByteChannel;
+
+import org.apache.spark.annotation.Private;
+
+/**
+ * :: Private ::
+ *
+ * A thin wrapper around a {@link WritableByteChannel}.
+ * <p>
+ * This is primarily provided for the local disk shuffle implementation to provide a
+ * {@link java.nio.channels.FileChannel} that keeps the channel open across partition writes.
+ *
+ * @since 3.0.0
+ */
+@Private
+public interface WritableByteChannelWrapper extends Closeable {
```

Comment:
why do we only need a wrapper for `WritableByteChannel`, but not `OutputStream`?

Reply:
We need to return the `FileChannel` object directly to the caller, because `FileChannel#transfer[from|to]` checks `instanceof` on the argument channel to transfer to/from in order to decide to optimize via zero-memory copy. Extending `FileChannel` is nearly impossible since it's an internal JDK abstract class with a lot of methods. But if we return the `FileChannel`, we have no way to shield the channel from being closed so that we can share the same channel resource across partitions.

This has come up in https://github.com/apache/spark/pull/25007#discussion_r306936154 and https://github.com/palantir/spark/pull/535 and especially https://github.com/palantir/spark/pull/535#discussion_r280193123. Given that this has come up as a question a number of times, I wonder if there's a better way we can make the semantics more accessible. I don't see a way to improve the architecture itself, but perhaps better documentation in the right places explaining why we went about this the way we did is warranted.